pkg/kube-metrics/collector.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2019 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubemetrics
+
+import (
+	"context"
+
+	"github.com/operator-framework/operator-sdk/pkg/k8sutil"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+	kcollector "k8s.io/kube-state-metrics/pkg/collector"
+	"k8s.io/kube-state-metrics/pkg/metric"
+	metricsstore "k8s.io/kube-state-metrics/pkg/metrics_store"
+)
+
+// NewCollectors returns collections of metrics in the namespaces provided, per the api/kind resource.
+// The metrics are registered in the custom generateStore function that needs to be defined.
+// Current operators namespace will be added to the passed namespaces if it is not present.
+func NewCollectors(uc *Client, namespaces []string, api string, kind string, metricFamily []metric.FamilyGenerator) ([]*kcollector.Collector, error) {
+	// Detect in which namespace the operator is deployed in.
+	operatorNamespace, err := k8sutil.GetOperatorNamespace()
+	if err != nil {
+		return nil, err
+	}
+	namespaces = append(namespaces, operatorNamespace)
+	namespaces = deduplicateNamespaces(namespaces)
+	var collectors []*kcollector.Collector
+	// Generate collector per namespace.
+	for _, ns := range namespaces {
+		dclient, err := uc.ClientFor(api, kind, ns)
+		if err != nil {
+			return nil, err
+		}
+		composedMetricGenFuncs := metric.ComposeMetricGenFuncs(metricFamily)
+		headers := metric.ExtractMetricFamilyHeaders(metricFamily)
+		store := metricsstore.NewMetricsStore(headers, composedMetricGenFuncs)
```

Comment:
We already have a cache for this, and now we are creating another. I think for now this makes sense as we may need upstream changes to share the same store, but WDYT of investigating that in the future?

Reply:
Right now not sure, as each one is bound to the metrics family that we generate and that is generated for each of them based on the custom resource name. But I will try to investigate if this is possible to make changes upstream. Will bring up the idea to the upstream devs!